File: simcore/framework/examples/my_module.py
#备注: python语言算法接入示例

import struct
import argparse
from txsim_module import *

# tcp topic例子
TEST_TCP_TOPIC = 'py_topic'
# 共享内存topic例子
TEST_SHMEM_TOPIC = 'py_topic_shmem'
# 备注:
# 参数:
# 返回值:


class MyModule(SimModule):
    def __init__(self):
        print('my Python module constructed.')
    # 备注: 该方法在算法第一次连接仿真系统时候被回调
    # 参数:
    #    helper: 参考算法接入文档
    # 返回值:
    def on_init(self, helper):
        print("========== my module on_init called. ==========")
        v = helper.get_parameter('max_step')
        self.max_step_time = 0
        if (len(v) > 0):
            self.max_step_time = int(v) * 1000
            print("max step time = {}s".format(v))
        self.pub = bool(helper.get_parameter('pub'))
        if self.pub:
            print("I'm a msg publisher.")
            helper.publish(TEST_TCP_TOPIC)
            helper.publish_shmem(TEST_SHMEM_TOPIC, 512)
        else:
            print("I'm a msg subscriber.")
            helper.subscribe(TEST_TCP_TOPIC)
            helper.subscribe_shmem(TEST_SHMEM_TOPIC)
    # 备注: 该方法在算法重置场景的时候被回调
    # 参数:
    #    helper: 参考算法接入文档
    # 返回值:
    def on_reset(self, helper):
        print("========== my module on_reset called. ==========")
        print("map file path: {}".format(helper.map_file_path()))
        lon, lat, alt = helper.map_local_origin()
        print("the map local origin is ({}, {}, {})".format(lon, lat, alt))
        lon, lat, alt = helper.ego_destination()
        print("the destination of the ego car is ({}, {}, {})".format(lon, lat, alt))
        print("speed limit: {}".format(helper.ego_speed_limit()))
        print("geo fence: {}".format(helper.geo_fence()))
        print("scenario file path: {}".format(helper.scenario_file_path()))
        print("ego path: {}".format(helper.ego_path()))
        vms = helper.vehicle_measurements()
        print("got {} vehicle measurements.".format(len(vms)))
        if len(vms) > 0:
            print("vehicle measurement: <id->{}, size->{}>".format(vms[0][0], len(vms[0][1])))
    # 备注: 该方法在仿真的时候被回调
    # 参数:
    #    helper: 参考算法接入文档
    # 返回值:
    def on_step(self, helper):
        print("========== my module on_step called. =========")
        t = helper.timestamp()
        print("current simulation timestamp: {}ms".format(t))
        if self.pub:
            msg = 'msg of py pub module on {} ms.'.format(t)
            helper.publish_message(TEST_TCP_TOPIC, msg.encode('utf-8'))
            buf = helper.get_published_shmem_buffer(TEST_SHMEM_TOPIC)
            if buf:
                print("got shared memory buffer with size of {} bytes.".format(len(buf)))
                struct.pack_into('I{}s'.format(len(msg)), buf, 0,
                                 len(msg), msg.encode('utf-8'))
        else:
            msg = helper.get_subscribed_message(TEST_TCP_TOPIC)
            print('received {} msg: {}'.format(
                TEST_TCP_TOPIC, msg.decode('utf-8')))
            buf = helper.get_subscribed_shmem_data(TEST_SHMEM_TOPIC)
            if buf:
                msg_size = struct.unpack_from('I', buf, 0)[0]
                print("found {} bytes data in shared memory.".format(msg_size))
                shmem_data = struct.unpack_from(
                    '{}s'.format(msg_size), buf, 4)[0]
                print("data in shared memory: {}".format(shmem_data))
        if self.max_step_time > 0 and t > self.max_step_time:
            helper.stop_scenario("max step time reached!")

    # 备注: 该方法在结束仿真系统时候被回调
    # 参数:
    #    helper: 参考算法接入文档
    # 返回值:
    def on_stop(self, helper):
        print("========== my module on_stop called. ==========")
        helper.set_feedback("my_key", "my_value")


# 备注: python入口函数
# 参数:
#    helper: 参考算法接入文档
# 返回值:
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--name', required=True)
    args = parser.parse_args()
    m = MyModule()
    s = SimModuleService()
    s.serve(args.name, m)
    print("python sim module service exit.")

File: simcore/framework/examples/my_module/my_module.cc
#include "my_module.h"

#include <iomanip>
#include <iostream>

// since the example would process the two types of message, 
// we need include the two corresponding protobuf headers here.
// if you need process other type of protobuf message that TADSim defines, include them too.
#include "traffic.pb.h"
#include "location.pb.h"
#include "settings.pb.h"
#include "scene.pb.h"


#ifdef _WIN32
#define MODULE_API __declspec(dllexport)
#else
#define MODULE_API
#endif // _WIN32

#define SPLIT_LINE "======================================="
#define SHMEM_TOPIC "MY_SHMEM_TOPIC"


MyModule::MyModule() {}


MyModule::~MyModule() {
  std::cout << "MyModule destroyed." << std::endl;
}


void MyModule::Init(tx_sim::InitHelper& helper) {
  std::cout << SPLIT_LINE << std::endl;

  // get user defined initiation parameters.
  // if we defined the parameters in TADSim UI, override the default values here.
  std::string value = helper.GetParameter("max_step_count");
  if (!value.empty()) {
    max_step_count_ = std::stoi(value);
    std::cout << "Init with parameter max_step_count = " << max_step_count_ << std::endl;
  }
  value = helper.GetParameter("velocity");
  if (!value.empty()) {
    step_velocity_ = std::stod(value);
    std::cout << "Init with parameter velocity = " << step_velocity_ << std::endl;
  }

  puber_ = !helper.GetParameter("pub").empty();
  if (puber_) {
    // publish our topics with messages we produced in Step callback.
    helper.Publish(tx_sim::topic::kLocation);
    // now we could also pub-subs through shared memory!
    helper.PublishShmem(SHMEM_TOPIC, 1024);
    // also subscribe location sent by itself.
    helper.Subscribe(tx_sim::topic::kLocation);
    //helper.ImmediateSubcribe(tx_sim::topic::kLocation);
  } else {
    // by subscribe our interested topics, we expect that the two corresponding messages which defined by
    // traffic.proto and location.proto would received in every Step callback.
    helper.Subscribe(tx_sim::topic::kTraffic);
    helper.Subscribe(tx_sim::topic::kLocation);
    // now we could also pub-subs through shared memory!
    helper.SubscribeShmem(SHMEM_TOPIC);
  }
  helper.Subscribe(tx_sim::topic::kUnionFlag + tx_sim::topic::kLocation);
};


void MyModule::Reset(tx_sim::ResetHelper& helper) {
  std::cout << SPLIT_LINE << std::endl;

  // first we should reset(clear) any data status before starting a new scenario.
  step_count_ = 0;
  last_timestamp_ = 0;
  start_x_ = 0, start_y_ = 0, cur_x_ = 0, cur_y_ = 0;

  // here we could get some information(e.g. the map which the scenario runs on, suggested local origin coordinate
  // and the destination of the ego car) of the current scenario.
  tx_sim::Vector3d o = helper.map_local_origin(), d = helper.ego_destination();
  sim_msg::Location start_location;
  // get start location in reset or in first step. both works.
  start_location.ParseFromString(helper.ego_start_location());
  std::cout << "Reset scenario with parameters:\n"
    << "hadmap path: " << helper.map_file_path() << "\n"
    << "the local origin coordinate of the map: (" << o.x << ", " << o.y << ", " << o.z << ")" << "\n"
    << "ego car's destination (" << d.x << ", " << d.y << ")" << std::endl;
  // we can also get the speed limit(max) info of the ego car in current scenario.
  double speed_limit = helper.ego_speed_limit();
  std::string path = helper.map_file_path();
  //helper.txsim_scence_events();
  // we could throw an exception which must inherited from std::exception type, if there is an error.
  if (helper.map_file_path() == "/some/map/that/we/cannot/load") {
    // std::runtime_error is a subclass of the std::exception.
    throw std::runtime_error("the map " + helper.map_file_path() + " could not load.");
  }

  //helper.setting_pb();
  sim_msg::Setting setting;
  std::string sSetting = helper.setting_pb();
  setting.ParseFromString(sSetting);
  std::cout << setting.DebugString()<<std::endl;

  //helper.scene_pb();
  sim_msg::Scene scene;
  std::string sScene = helper.scene_pb();
  scene.ParseFromString(sScene);
  std::cout << scene.DebugString()<<std::endl;

};


void MyModule::Step(tx_sim::StepHelper& helper) {
  std::cout << SPLIT_LINE << std::endl;

  // 1. get current simulation timestamp.
  double time_stamp = helper.timestamp();
  std::cout << "time stamp: " << time_stamp << "\n";
  int64_t curtime_stamp = helper.current_timestamp();
  //std::cout << "curtime_stamp : " << curtime_stamp << "\n";

  // 2. get messages we subscribed.
  helper.GetSubscribedMessage(tx_sim::topic::kLocation, payload_);
  sim_msg::Location loc;
  loc.ParseFromString(payload_);
  cur_x_ = loc.position().x();
  cur_y_ = loc.position().y();
  std::cout << std::fixed << std::setprecision(12) << "received location: x -> " \
   << cur_x_ << " y -> " << cur_y_ << std::endl;

  helper.GetSubscribedMessage(tx_sim::topic::kUnionFlag + tx_sim::topic::kLocation, payload_);
  std::cout << "union message length: " << payload_.size() << std::endl;
  if (puber_) {
    // 3. here should put the actual user algorithm, do some computing according to the subscribed messages we received.
    // for explanatory simplicity, it only moves a little by a constant velocity, no matter what happens.
    double move_distance = step_velocity_ * (time_stamp - last_timestamp_); // s = v * t
    std::cout << "ego car moved " << move_distance << std::endl;
    double next_x = cur_x_ + move_distance, next_y = cur_y_ + move_distance;

    // 4. put our results into output messages and publish them.
    loc.mutable_position()->set_x(next_x);
    loc.mutable_position()->set_y(next_y);
    // since the message we want to publish is a protobuf message type
    // and the PublishMessage API only accepts the std::string type,
    // we need serialize it to std::string using google::protobuf::MessageLite::SerializeToString method.
    loc.SerializeToString(&payload_);
    helper.PublishMessage(tx_sim::topic::kLocation, payload_);

    last_timestamp_ = time_stamp;
    std::cout << "expected next position: (" << next_x << ", " << next_y << ")" << std::endl;

    // we could also send data via shared memory ...
    char* shm_buf = nullptr;
    uint32_t shm_size = helper.GetPublishedShmemBuffer(SHMEM_TOPIC, &shm_buf);
    if (shm_buf == nullptr) {
      std::cerr << "no shm available!" << std::endl;
    } else {
      std::string shm_data = "this is a string data on step " + std::to_string(step_count_);
      size_t str_len = shm_data.size();
      memcpy(shm_buf, &str_len, sizeof(size_t));
      memcpy(shm_buf + sizeof(size_t), shm_data.data(), str_len);
    }
  } else {
    helper.GetSubscribedMessage(tx_sim::topic::kTraffic, payload_);
    // since the traffic and location message is defined by TADSim, the payload string we get here is
    // in serialized protobuf bytes form, we need deserialize them manually.
    // using google::protobuf::MessageLite::ParseFromString method to parse from a std::string type.
    sim_msg::Traffic traffic;
    traffic.ParseFromString(payload_);
    std::cout << tx_sim::topic::kTraffic << ": "
      << traffic.cars_size() << " cars, "
      << traffic.staticobstacles_size() << " static obstacles, "
      << traffic.dynamicobstacles_size() << " dynamic obstacles, "
      << traffic.trafficlights_size() << " traffic lights.\n";

    const char* shm_buf = nullptr;
    uint32_t shm_size = helper.GetSubscribedShmemData(SHMEM_TOPIC, &shm_buf);
    if (shm_buf == nullptr) {
      std::cerr << "no shm available!" << std::endl;
    } else {
      std::cout << "shm buf address: " << (void*)shm_buf << std::endl;
      size_t data_len = 0;
      memcpy(&data_len, shm_buf, sizeof(size_t));
      std::cout << "copied length from shm buf: " << data_len << std::endl;
      std::unique_ptr<char[]> read_data(new char[data_len + 1]);  // don't forget the last '\0'
      memcpy(read_data.get(), shm_buf + sizeof(size_t), data_len);
      read_data[data_len] = '\0';
      std::cout << "read data from shared memory: " << read_data.get() << std::endl;
    }
  }

  // this is just for simplicity. we should stop the scenario when we reached the destination point
  // which we received in Reset() by helper.ego_destination() method.
  step_count_++;
  if (step_count_ >= max_step_count_) {
    helper.StopScenario("we have reached our destination.");
  }
};


void MyModule::Stop(tx_sim::StopHelper& helper) {
  std::cout << SPLIT_LINE << std::endl;
  std::cout << name_ << " module stopped." << std::endl;
  // here we could send out any feedback data we want the TADSim to display on front-end UI.
  helper.set_feedback("stepCounts", std::to_string(step_count_));

  std::string str = helper.get_module_status();
  std::cout << " module status: " << str << std::endl;
};


TXSIM_MODULE(MyModule)

